src/migrations/merge.rs

```diff
@@ -0,0 +1,172 @@
+use indexmap::IndexMap;
+use fs_err as fs;
+
+use crate::connect::Connection;
+use crate::migrations::{Context, migrate, migration};
+use crate::migrations::create::{MigrationKey, MigrationToText, write_migration};
+use crate::migrations::db_migration::{DBMigration, read_all};
+use crate::migrations::migration::MigrationFile;
+use crate::migrations::rebase::{fix_migration_ids};
+use crate::print;
+
+pub struct MergeMigrations {
+    pub base_migrations: IndexMap<String, MergeMigration>,
+    pub target_migrations: IndexMap<String, MergeMigration>
+}
+
+impl MergeMigrations {
+    fn flatten(&self) -> IndexMap<&String, &MergeMigration> {
+        let mut result = IndexMap::new();
+
+        for migration in self.base_migrations.iter() {
+            result.insert(migration.0, migration.1);
+        }
+
+        for migration in self.target_migrations.iter() {
+            result.insert(migration.0, migration.1);
+        }
+
+        result
+    }
+}
+
+pub struct MergeMigration {
+    key: MigrationKey,
+    id_override: Option<String>,
+    migration: DBMigration,
+    parent_override: Option<String>
+}
+
+impl MigrationToText for MergeMigration {
+    type StatementsIter<'a> = std::iter::Once<&'a String> where Self: 'a;
+
+    fn key(&self) -> &MigrationKey {
+        &self.key
+    }
+
+    fn parent(&self) -> anyhow::Result<&str> {
+        Ok(self.parent_override.as_ref().or(self.migration.parent_names.first()).map(|v| v.as_str()).unwrap_or("initial"))
+    }
+
+    fn id(&self) -> anyhow::Result<&str> {
+        Ok(self.id_override.as_ref().unwrap_or(&self.migration.name))
+    }
+
+    fn statements<'a>(&'a self) -> Self::StatementsIter<'a> {
+        std::iter::once(&self.migration.script)
+    }
+}
+
+pub async fn get_merge_migrations(base: &mut Connection, target: &mut Connection) -> anyhow::Result<MergeMigrations> {
+    let base_migrations = read_all(base, true, false).await?;
+    let mut target_migrations = read_all(target, true, false).await?;
+
+    if base_migrations.len() > target_migrations.len() {
+        anyhow::bail!("Source branch contains more migrations than the target branch");
+    }
```

Comment:
Can we drop this special message? I goes against my mental model of how merge works.

Reply:
Should the if case be removed entirely? or should the message just change

Reply:
I'd say remove the whole if